depends/libhdfs3/src/client/CryptoCodec.cpp

```diff
@@ -0,0 +1,163 @@
+/********************************************************************
+ * 2014 -
+ * open source under Apache License Version 2.0
+ ********************************************************************/
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "CryptoCodec.h"
+#include "Logger.h"
+
+using namespace Hdfs::Internal;
+
+namespace Hdfs {
+
+/**
+ * Construct a CryptoCodec instance.
+ * @param encryptionInfo the encryption info of file.
+ * @param kcp a KmsClientProvider instance to get key from kms server.
+ * @param bufSize crypto buffer size.
+ */
+CryptoCodec::CryptoCodec(FileEncryptionInfo *encryptionInfo, std::shared_ptr<KmsClientProvider> kcp, int32_t bufSize) : encryptionInfo(encryptionInfo), kcp(kcp), bufSize(bufSize)
+{
+		
+	/* Init global status. */
+	ERR_load_crypto_strings();
+	OpenSSL_add_all_algorithms();
+	OPENSSL_config(NULL);
+
+	/* Create cipher context. */
+	encryptCtx = EVP_CIPHER_CTX_new();		
+	cipher = NULL;	
+
+}
+
+/**
+ * Destroy a CryptoCodec instance.
+ */
+CryptoCodec::~CryptoCodec()
+{
+	if (encryptCtx) 
+		EVP_CIPHER_CTX_free(encryptCtx);
+}
+
+/**
+ * Get decrypted key from kms.
+ */
+std::string CryptoCodec::getDecryptedKeyFromKms()
+{
+	ptree map = kcp->decryptEncryptedKey(*encryptionInfo);
+	std::string key = map.get<std::string>("material");
+
+	int rem = key.length() % 4;
+	if (rem) {
+		rem = 4 - rem;
+		while (rem != 0) {
+			key = key + "=";
+			rem--;
+		}
+	}
+
+	std::replace(key.begin(), key.end(), '-', '+');
+	std::replace(key.begin(), key.end(), '_', '/');
+
+	LOG(INFO, "material is :%s", key.c_str());
```

Comment:
Suggest provide more clear log message, and if this function is called frequently, use DEBUG3 instead of INFO.